### ISSUE TITLE
Support rules from flake8-force-keyword-arguments

### ISSUE THREAD

**Issue by Riezebos** [astral-sh/ruff#3269]
Calling a function with a lot of unnamed arguments is a lot less readable than using keyword arguments. These rules enforce keyword arguments if a function has a lot of parameters, which should lead to more readable code. I believe adding them to ruff would be a good idea!

https://github.com/isac322/flake8-force-keyword-arguments

If there is anything I can do for this as a Python dev with practically 0 Rust experience, I would be happy to help.

**Comment by Riezebos** [comment]
Thanks for the response! I think it would be partially possible with only local information:

- Rule 1: Flag any function definition that accepts >X positional arguments
- Rule 2: Flag any function call that uses >X positional arguments

For the first rule I think only local information is needed? An autofix could potentially be provided that adds `*, ` at the start of the parameter definition (and removes any other `*,` or `/,` from the definition).

With only local information rule 2 would falsely flag function calls of functions with >X positional-only parameters or with *args. I would personally be fine with manually adding noqa comments there but I can imagine that this disqualifies it.

Would rule 1 be an option?

**Comment by flying-sheep** [comment]
We would definitely benefit from a rule that flags function *definitions* that use too many positional arguments.

[PLR0913](https://docs.astral.sh/ruff/rules/too-many-arguments/) is about too many total arguments in function definitions, which is something we can’t really change in our API.

But we *can* add `*`s to make our API better. And a rule checking this would help a lot.

**Comment by emilte** [comment]
I would love this.
I currently have a custom implementation of this in my project with pylint.
I'm considering to replace everything with ruff, but will surely miss this feature.
A star * at the beginning of functions helps immensely

**Comment by flying-sheep** [comment]
`R0917` aka “too-many-positional” is implemented in Ruff already.

**Comment by emilte** [comment]
Aha, that's great, thx!

I struggled a bit to get this working, is there a tiny bug in the documentation?
The example shows `max-pos-args`, but `max-positional-args` seems to be the correct key.
https://docs.astral.sh/ruff/settings/#pylint-max-positional-args

**Comment by dhruvmanila** [comment]
> The example shows `max-pos-args`, but `max-positional-args` seems to be the correct key.

Thanks! It must be a typo. Do you want to open a PR fixing the docs? The type is here:

https://github.com/astral-sh/ruff/blob/467c091382fa60243e261417b4620aeca4012bf5/crates/ruff_workspace/src/options.rs#L2766

**Comment by emilte** [comment]
Thx for the opportunity.
https://github.com/astral-sh/ruff/pull/9797

**Comment by tmke8** [comment]
I think this issue can be closed then, since it's possible to achieve the desired behavior with `PLR0917` and setting

```toml
[tool.ruff.lint.pylint]
max-positional-args = 0
```

**Comment by jack-mcivor** [comment]
@Riezebos are you sure this is closed? 

> Rule 1: Flag any function definition that accepts >X positional arguments
> Rule 2: Flag any function call that uses >X positional arguments

AFAICT PLR0917 checks function definitions only (Rule 1). I don't think a check for function calls exists? (Rule 2)

**Comment by Andrew-S-Rosen** [comment]
Related, what would be super useful is some rule that would ensure that arguments are explicitly specified in keyword format when calling a function that accepts keyword arguments. The purpose here is that if the order or number of arguments in the function signature is ever updated, the user's code would remain unaffected. This would, however, require inspection of each function signature, including those of dependencies.

This might not be viable in the short-term, but I wanted to throw out additional motivation beyond "it just looks nicer."

**Comment by andrewemporia** [comment]
I would also really like Rule 2 as it makes this much easier to roll out to an existing codebase. If I turn on Rule 1 and someone touches a large file they will have to go add '*' to every large function definition in that file and **then** change every caller of those functions which is a large amount of work. Just requiring function calls to be keyword if passing over a certain number of args is much more palatable to introduce change for a team. And an autofix would be fantastic.

**Comment by toppk** [comment]
While I agree with @andrewemporia that making arguments keyword-only causes a big burden, that's sometimes the correct fix.  I'd like a rule that finds functions that have lots of identically typed keyword paramenters that aren't keyword-only.  that's the mistake I make.
```python
def foo(a:int,b:int, data1:dict[str,int]|None=None, data2:dict[str,int]|None=None) -> bool:  ...
```
I find too many times i'm calling 
```python
foo(1,2,{'hi':3})  # data1
```
 thinking that's `data2`, and not `data1`.  in this case the only fix is making data1/data2 kwonly, i.e.


```python
def foo(a:int,b:int,*, data1:dict[str,int]|None=None, data2:dict[str,int]|None=None) -> bool:  ...
```

so i'd like a rule that isn't about positional, but about similarly typed positional-or-keyword parameters.  once the parameters are keyword-only, my type checker (or runtime :-/ ) will spot that the callers are broken.  it would be incremental by fixing the caller's first before moving to keyword-only, but the proper mechanism for wanting keyword-only is using keyword-only, IMO.

**Comment by LeonardoRosaa** [comment]
PLR0917 (`too-many-positional-arguments`) solves this for me. It's definition-side, not call-site, but same effect in practice.

```toml
[tool.ruff.lint]
extend-select = ["PLR0917"]

[tool.ruff.lint.pylint]
max-positional-args = 1
```

```python
def foo(*, a: str) -> None: ...
```

With `*`, `foo("bar")` is now a `TypeError`, so callers have to do `foo(a="bar")`.
